data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostProposerSlashing.java

```diff
@@ -69,25 +92,27 @@ public PostProposerSlashing(final NodeDataProvider provider, final JsonProvider
         @OpenApiResponse(status = RES_INTERNAL_ERROR),
       })
   @Override
-  public void handle(final Context ctx) throws Exception {
-    try {
-      final ProposerSlashing proposerSlashing =
-          parseRequestBody(ctx.body(), ProposerSlashing.class);
-      InternalValidationResult result =
-          nodeDataProvider.postProposerSlashing(proposerSlashing).join();
-      if (result.code().equals(ValidationResultCode.IGNORE)
-          || result.code().equals(ValidationResultCode.REJECT)) {
-        ctx.status(SC_BAD_REQUEST);
-        ctx.json(
-            BadRequest.badRequest(
-                jsonProvider,
-                "Invalid proposer slashing, it will never pass validation so it's rejected"));
-      } else {
-        ctx.status(SC_OK);
-      }
-    } catch (final IllegalArgumentException | JsonMappingException e) {
-      ctx.json(BadRequest.badRequest(jsonProvider, e.getMessage()));
-      ctx.status(SC_BAD_REQUEST);
-    }
+  public void handle(@NotNull final Context ctx) throws Exception {
+    adapt(ctx);
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final ProposerSlashing proposerSlashing = request.getRequestBody();
+    final SafeFuture<InternalValidationResult> future =
+        nodeDataProvider.postProposerSlashing(proposerSlashing);
+
+    request.respondAsync(
+        future.thenApply(
+            internalValidationResult -> {
+              if (internalValidationResult.code().equals(ValidationResultCode.IGNORE)
+                  || internalValidationResult.code().equals(ValidationResultCode.REJECT)) {
+                return AsyncApiResponse.respondWithError(
+                    SC_BAD_REQUEST,
+                    "Invalid proposer slashing, it will never pass validation so it's rejected");
```

Comment:
same here i think , can use internalValidationResult description or fall back to generic message if its not set...